### PR TITLE
Division with timedelta input

### PR DESCRIPTION
**Pull request by luk-f-a**
This PR enables the division of a timedelta array by a scalar or array of a different type.

The main change is typing of the output. Currently  `timedelta64[M] / int` is incorrecly typed as `timedelta64[]` which produces as a lowering error `No conversion from timedelta64[M] to timedelta64[]`.
These changes make the output to be typed as `timedelta64[M]` (or any other time unit that matches the input), which eliminates the lowering error.

This PR builds on #5060 (and includes the changes from that PR) and is required before #4480.

### CODE REVIEW

**Comment by gmarkall** [comment]
Thanks very much for the PR, I've queued it for review.

File: numba/tests/test_array_exprs.py [hunk 1]
```diff
@@ -613,5 +615,21 @@ def do_call(x, y):
...
+            return a / b
+
+        py_func = arr_div
+        cfunc = njit(arr_div)
+        test_cases = [(np.ones(3, np.dtype(TIMEDELTA_M)), np.ones(3, np.dtype(TIMEDELTA_M))),
+                      (np.ones(3, np.dtype(TIMEDELTA_Y)), np.ones(3, np.dtype(TIMEDELTA_Y))),
+                      (np.ones(3, np.dtype(TIMEDELTA_M)), 1),
+                      (np.ones(3, np.dtype(TIMEDELTA_M)), np.ones(3, np.int64)),
+                      (np.ones(3, np.dtype(TIMEDELTA_M)), np.ones(3, np.float64)),
+                     ]
```
Comment (guilhermeleobas) [inline numba/tests/test_array_exprs.py:633]: Test the case where `a` have dtype `TIMEDELTA_M` and `b` have dtype `TIMEDELTA_Y`:
```python
(np.ones(3, np.dtype(TIMEDELTA_M)), np.ones(3, np.dtype(TIMEDELTA_Y))),`
```

File: numba/np/numpy_support.py [hunk 2]
```diff
@@ -391,25 +391,46 @@ def set_output_dt_units(inputs, outputs):
...
 
-        At the moment this function takes care of the case where all
-        inputs have the same unit, and therefore the output unit has the same.
-        If in the future this should be extended to a case with mixed units,
-        the rules should be implemented in `npdatetime_helpers` and called
-        from this function to set the correct output unit.
+        At the moment this function takes care of two cases
+         a) where all inputs are timedelta with the same unit (mm), and
+         therefore the output has the same unit.
+         If in the future this should be extended to a case with mixed units,
```
Comment (guilhermeleobas) [inline numba/np/numpy_support.py:397]: The `If` in this phrase is correct? I think it should be `In the future this should be extended ...`?

File: numba/tests/test_array_exprs.py (hunk 1 above)
Comment (luk-f-a) [reply numba/tests/test_array_exprs.py]: thanks, I thought numpy would reject that operation, it actually works. Added two tests.

File: numba/np/numpy_support.py (hunk 2 above)
Comment (luk-f-a) [reply numba/np/numpy_support.py]: thanks, you are right, it was wrong. I kept the if, but changed the should.

**Comment by luk-f-a** [comment]
@guilhermeleobas thanks for the review. I addressed your points in the latest commits.

**Comment by esc** [comment]
@guilhermeleobas thanks for reviewing this. We are planning to release a 0.50.0RC1 this week so this will likely be parked until after that since it is in the 0.51.0 milestone. I'll mark it as 'waiting for reviewer' since an initial review has taken place and to attract attention .

File: numba/tests/test_array_exprs.py [hunk 3]
```diff
@@ -613,5 +614,24 @@ def do_call(x, y):
         self.assertTrue(isinstance(oty.type.dtype, types.Float))
 
 
-if __name__ == "__main__":
-    unittest.main()
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/tests/test_array_exprs.py:617]: oops :-/ I don't know how it happened. Fixed it!

File: numba/np/numpy_support.py [hunk 4]
```diff
@@ -391,25 +391,46 @@ def set_output_dt_units(inputs, outputs):
...
+         If in the future this needs to be extended to a case with mixed units,
+         the rules should be implemented in `npdatetime_helpers` and called
+         from this function to set the correct output unit.
+         b) where left operand is a timedelta, ie the "m?" case. This case
+         is used for division, eg timedelta / int.
+
+         At the time of writing, numba does not support addition of timedelta
+         and other types, so this function does not consider the case "?m",
+         ie where timedelta is the right operand to a non-timedelta left operand
+         To extend it in the future, just add another elif clause.
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/np/numpy_support.py]: fixed

**Comment by stuartarchibald** [comment]
Flake8 is complaining about a long line in the branch HEAD:
```
numba/np/numpy_support.py:407:81: E501 line too long (81 > 80 characters)
```

File: numba/np/numpy_support.py [hunk 5]
```diff
@@ -391,25 +391,46 @@ def set_output_dt_units(inputs, outputs):
...
+         are timedeltas
+         If in the future this needs to be extended to a case with mixed units,
+         the rules should be implemented in `npdatetime_helpers` and called
+         from this function to set the correct output unit.
+         b) where left operand is a timedelta, ie the "m?" case. This case
+         is used for division, eg timedelta / int.
+
+         At the time of writing, numba does not support addition of timedelta
+         and other types, so this function does not consider the case "?m",
+         ie where timedelta is the right operand to a non-timedelta left operand
```
[the remark this answers is not in the document]
Comment (luk-f-a) [reply numba/np/numpy_support.py]: `numba/np/numpy_support.py:407:81: E501 line too long (81 > 80 characters)` :rofl:

**Pull request merged**